### PR TITLE
Bootstrap practice exercises

### PR DESCRIPTION
**Pull request by jiegillet**
This script will generate all files (minus config.json and tests.toml) for a practice exercise. Of course, the tests need a bunch of manual input, but usually at the level of search and replace, so it's quite valuable.

I tried to be a little bit clever about some things (for example @spec generation for the lib file) but I hope that contributors understand that they should not trust these guesses too much. Is it better to make them obviously wrong?
Please test that script by running `elixir bin/boostrap_practice_exercise.exs darts` (or whichever exercise) and inspect the generated files. I'm sure there is room for improvement.

### CODE REVIEW

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [inline bin/bootstrap_practice_exercise.exs:44]: > Is it better to make them obviously wrong?

Maybe it's better to somehow mark them as requiring manual changes? Maybe something like this? (and similar in other places that also have most-likely-wrong defaults)
```suggestion
    """
     @doc \"\"\"
     TODO: insert function description here
     \"\"\"
     @spec #{property}(#{Enum.map_join(variables, ", ", fn
      {var, nil} -> "#{var} :: TODO.t()"
      {var, sub_vars} -> "#{var} :: %{#{Enum.map_join(sub_vars, " : TODO.t(), ", fn v -> v end)} : TODO.t()}"
    end)}) :: TODO.t()
     def #{property}(#{Enum.map_join(variables, ", ", fn {var, _} -> var end)}) do
     end
    """
```

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [inline bin/bootstrap_practice_exercise.exs:99]: It would be good if changing to snake case happened as soon as possible, right after fetching and decoding the JSON, and not when printing the test case. For example, you forgot to use it when printing the stub 😉 so for an exercise like `allergies` (https://github.com/exercism/problem-specifications/blob/e2779fc84ff07cb5bd8ff89f38e78af6d17b5056/exercises/allergies/canonical-data.json#L14), you get a stub like this:
```elixir
def allergicTo(item, score) do
end
```

but a function call in tests like this:
```elixir
output = Allergies.allergic_to(item, score)
```

It's also necessary to change properties to snake case, for example: https://github.com/exercism/problem-specifications/blob/daf620d47ed905409564dec5fa9610664e294bde/exercises/all-your-base/canonical-data.json#L20-L22

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [inline bin/bootstrap_practice_exercise.exs:74]: Would it be possible (and not too difficult) to make this more intelligent, and only use `{:ok, value}` if any error case exists in the problem specification for this property, and otherwise just use the plain value? There are a lot of exercises without any error handling.

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [inline bin/bootstrap_practice_exercise.exs:140]: I had no idea about this module 😍 awesome

**Review by angelikatyborska — APPROVED** [review]
I run it on a few existing exercises and left some (hopefully realistic) improvement ideas based on the output.

This is excellent 🎉. I guess with this script you'll churn through all the remaining "new practice exercise" issues in no time. I have to create more 😅.

**Comment by neenjaw** [comment]
I'd like to look over this one before merge, but might need to wait a couple days if that's okay.  I have all next week off, so should be able to catch up on this stuff. :)

File: bin/bootstrap_practice_exercise.exs
Comment (jiegillet) [reply bin/bootstrap_practice_exercise.exs:140]: It is quite nice. And actually it wouldn't be much more work to make this script into a `Mix.Task`. It would be kind of cool: `mix ex.practice.gen darts`.

File: bin/bootstrap_practice_exercise.exs
Comment (jiegillet) [reply bin/bootstrap_practice_exercise.exs]: Done :)

File: bin/bootstrap_practice_exercise.exs
Comment (jiegillet) [inline bin/bootstrap_practice_exercise.exs:282]: Do you know if there is a way to call `mix format` from within a `.exs` script without having to use `System.cmd`? I searched for it, but couldn't find it.  
Although maybe it's best to run it this way so that it includes the  format in `.formatter.exs`.

Comment (angelikatyborska) [reply bin/bootstrap_practice_exercise.exs:282]: Every mix task is just a module with a `run` function, so it should be possible to run `Mix.Tasks.Format.run([])`. It should read `.formatter.exs` from the current directory anyway.

mix format documentation: https://hexdocs.pm/mix/master/Mix.Tasks.Format.html
and implementation: https://github.com/elixir-lang/elixir/blob/2a4312412be39303af08ccd0cef81b5bbcd55eac/lib/mix/lib/mix/tasks/format.ex#L1

**Comment by jiegillet** [comment]
> I guess with this script you'll churn through all the remaining "new practice exercise" issues in no time. I have to create more 😅.

I compiled a list of all the exercises we don't have yet (separating the ones with and without tests), do you want me to share it in an issue? Or do you prefer individual ones?

File: bin/bootstrap_practice_exercise.exs
Comment (neenjaw) [inline bin/bootstrap_practice_exercise.exs:13]: We can use `Macro.underscore` for this if we are using it just for identifiers

File: bin/bootstrap_practice_exercise.exs
Comment (neenjaw) [inline bin/bootstrap_practice_exercise.exs:133]: I don't know if there is really a standard established for mix install, but can we hoist this to the top of the file?

File: bin/bootstrap_practice_exercise.exs
Comment (neenjaw) [inline bin/bootstrap_practice_exercise.exs:194]: Minor: if we can't run mix format on this file, can we indent here 2 spaces for this binding?

File: bin/bootstrap_practice_exercise.exs
Comment (neenjaw) [inline bin/bootstrap_practice_exercise.exs:202]: Minor, same indent as mentioned above

**Review by neenjaw — APPROVED** [review]
Left a few minor comments, cool stuff!

**Comment by angelikatyborska** [comment]
> I compiled a list of all the exercises we don't have yet (separating the ones with and without tests), do you want me to share it in an issue? Or do you prefer individual ones?

There is a dashboard for this already: https://tracks.exercism.io/elixir/main/unimplemented

Sadly it doesn't mention which exercises don't have tests defined, but there's only a few left unimplemented so we can manage.

Normally only maintainers deal with exercise selection, but since you're working faster than I can review or write new issues, I want to share with you my thoughts on implementing those missing exercises:

1. I don't want any of the exercises that don't have `canonical-data.json` (error handling, hangman, ledger, lens-person, paasio, robot-name, tree building) because that puts on me and Tim the responsibility to figure out which tests cases are necessary for the story, to think of the corner cases etc., it just sounds like too much work 😅 
2. I don't want any of the exercises that are ridiculously easy in a high level programming language (reverse string, maybe micro blog). That might not apply to square root because I think square root could forbid using the standard library and demand a manual implementation.
3. I don't want high scores because it will conflict with a similar concept exercise that we already have

The rest can be implemented and your PRs with them will be very welcome if you still have the time and energy 🙂

File: bin/bootstrap_practice_exercise.exs
Comment (jiegillet) [reply bin/bootstrap_practice_exercise.exs:194]: At the end we run mix format on everything, it should take care of this one too.

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [inline bin/bootstrap_practice_exercise.exs:188]: I think this is what @neenjaw means:
```suggestion
mix = """
defmodule #{module}.MixProject do
  use Mix.Project

  def project do
    [
      app: :#{exercise_snake_case},
      version: "0.1.0",
      # elixir: "~> 1.8",
      start_permanent: Mix.env() == :prod,
      deps: deps()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
    ]
  end
end
"""
```

File: bin/bootstrap_practice_exercise.exs
Comment (angelikatyborska) [reply bin/bootstrap_practice_exercise.exs:196]: Here the indentation looks good already 🤔 meaning: level 0, there's no indentation.

File: bin/bootstrap_practice_exercise.exs
Comment (jiegillet) [reply bin/bootstrap_practice_exercise.exs]: How obvious. Thank you :)

**Comment by angelikatyborska** [comment]
I'm going to make a call that all the feedback has been considered and applied, and now it's merge time 🎉

**Pull request merged**
